core/src/main/scala/org/apache/spark/status/AppStatusListener.scala

```diff
@@ -645,8 +645,11 @@ private[spark] class AppStatusListener(
   }
 
   override def onTaskEnd(event: SparkListenerTaskEnd): Unit = {
-    // TODO: can this really happen?
-    if (event.taskInfo == null) {
+    // TODO: can taskInfo null really happen?
+    // For resubmitted tasks caused by ExecutorLost, the SparkListenerTaskEnd is useless and
+    // will make activeTask in stage to be negative, this will cause stage not be removed in
+    // liveStages, and finally cause executor not removed in deadExecutors
+    if (event.taskInfo == null ||  event.reason == Resubmitted) {
```

Comment:
to help understand the code more, do you know when/where the scheduler handles resubmitted tasks?

Reply:
1. First in TaskSetManager.scala, we handle the executorLost
``` scala
  override def executorLost(execId: String, host: String, reason: ExecutorLossReason): Unit = {
    // Re-enqueue any tasks that ran on the failed executor if this is a shuffle map stage,
    // and we are not using an external shuffle server which could serve the shuffle outputs.
    // The reason is the next stage wouldn't be able to fetch the data from this dead executor
    // so we would need to rerun these tasks on other executors.
    if (isShuffleMapTasks && !env.blockManager.externalShuffleServiceEnabled && !isZombie) {
      for ((tid, info) <- taskInfos if info.executorId == execId) {
        val index = info.index
        // We may have a running task whose partition has been marked as successful,
        // this partition has another task completed in another stage attempt.
        // We treat it as a running task and will call handleFailedTask later.
        if (successful(index) && !info.running && !killedByOtherAttempt.contains(tid)) {
          successful(index) = false
          copiesRunning(index) -= 1
          tasksSuccessful -= 1
          addPendingTask(index)
          // Tell the DAGScheduler that this task was resubmitted so that it doesn't think our
          // stage finishes when a total of tasks.size tasks finish.
          sched.dagScheduler.taskEnded(
            tasks(index), Resubmitted, null, Seq.empty, Array.empty, info)
        }
      }
    }
```
2. the successful task is added to queue to re-execute in `addPendingTask`
3. `sched.dagScheduler.taskEnded` is called to tell DAGScheduler to handle this, this will post a `CompletionEvent` to eventProcessLoop  
4. the function `handleTaskCompletion` in DATScheduler.scala will handle this message, as the reason `Resubmitted` is kind of a failure, it will call the function `handleResubmittedFailure`,  which add the task partitionId to ShuffleMapStage's pendingPartitions. Thus this stage will wait this resubmitted task to finish again.

Reply:
Couple of things to clarify (the pr description/comments confused me a bit).

a) There was a pair of task start and task end event which were fired for the task (let us call it Tr)
b) When executor which ran Tr was lost, while stage is still running, a `Resubmitted` is fired for Tr.
c) Subsequently, a new task start and task end will be fired for the retry of Tr.


The issue here is, (b) is associated with (a) - which was a success earlier, but has now been marked as a failure.


We should not be ignoring this event, but rather deal with it properly.
For example, update counters, number of failed tasks, etc.

Reply:
@cloud-fan @Ngone51 @mridulm any advice for the metrics and counters?

Reply:
@wineternity  Thanks for the detailed analysis. Seems like the only metric that the `Resubmmitted` task end event could affect is the failed/completed tasks counters. And other metrics like executor summary, and shuffle read/write metrics are duplicated. So I tend to agree to recognize the `Resubmmitted` task end event as a signal rather than a real task end event. 

@mridulm WDYT?

Reply:
It has been a while since I looked at it, but this is what I expect to happen.

- When a task is marked a `Resubmitted`, the stage metrics are reverted from the earlier (successful) task completion. (`metricsDelta`)
- Failed tasks counter should go up - for both stage and executor.
- Active tasks MUST not go down - this is the issue we have to fix.

Rest of the method should not have too much of a change imo.

Cab the change not simply be limited to `if (event.reason != Resubmitted) <foo>.activeTasks -= 1` ?

Reply:
> Can the change not simply be limited to...

Ok..if our target is to only be able to remove the dead executors, I think this's good enough. But if we also want to correct the metrics, it's not enough.